usecases/modulecomponents/batch/batch.go

```diff
@@ -124,108 +159,210 @@ func (b *Batch) batchWorker() {
 			objCounter++
 		}
 
-		for objCounter < len(job.texts) {
-			if job.ctx.Err() != nil {
-				for j := objCounter; j < len(job.texts); j++ {
-					if !job.skipObject[j] {
-						job.errs[j] = fmt.Errorf("context deadline exceeded or cancelled")
-					}
-				}
+		// if we have a high rate limit we can send multiple batches in parallel.
+		//
+		// If the rate limit is high enough to "fit" the current batch, we send it concurrently. If not, we wait for
+		// either
+		//   - the rate limit to refresh, so we can schedule another concurrent batch
+		//   - the current batch to finish, so the next batch can be sent sequentially
+		//
+		// While using the same code both modes are working slightly different:
+		// 	- For concurrent batching, the amount of used tokens/requests is reserved as long as the batch is running
+		//	  and is cleared when it finishes. This ensures that we never exceed the rate limit and don't need to check
+		//	  the rate limit in the sendBatch function (we use a dummy that never fails a check). All updates happen
+		//	  in the main loop.
+		//  - For sequential batching, the rate limit  will be passed into the sendBatch function and is observed and
+		//    updated there. This allows to use the rate-limit in an optimal way, but also requires more checks. No
+		//    concurrent batch can be started while a sequential batch is running.
+
+		for {
+			timePerToken, objectsPerBatch = b.updateState(rateLimitPerApiKey, timePerToken, objectsPerBatch)
+			numRequests := 1 + int(1.25*float32(len(job.texts)))/objectsPerBatch // round up to be on the safe side
+			if rateLimit.CanSendFullBatch(numRequests, job.tokenSum) {
+				jobCopy := job.copy()
+				rateLimit.ReservedRequests += numRequests
+				rateLimit.ReservedTokens += job.tokenSum
+				timePerToken := timePerToken
+				numRequests := numRequests
```

Comment:
I don't think it is necessary, both of those variables are already set in a loop

Reply:
Annoyingly it is necessary, had tests failing with data-races without this.

The outer loop can modify these values through `b.updateState` while the goroutine is accessing them